src/main/java/ch/njol/skript/expressions/ExprIndices.java

```diff
@@ -18,81 +18,120 @@
 import ch.njol.skript.doc.Name;
 import ch.njol.skript.doc.Since;
 import ch.njol.skript.lang.Expression;
 import ch.njol.skript.lang.ExpressionType;
-import ch.njol.skript.lang.SkriptParser;
-import ch.njol.skript.lang.UnparsedLiteral;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
 import ch.njol.skript.lang.Variable;
 import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.skript.registrations.Comparators;
+import ch.njol.skript.util.LiteralUtils;
 import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.Map;
+import java.util.Map.Entry;
 
-@Name("Indexes of List")
-@Description("Returns all the indexes of a list variable")
-@Examples("set {l::*} to \"some\", \"cool\" and \"values\"\n" +
-		"broadcast \"%all indexes of {l::*}%\" # result is 1, 2 and 3")
-@Since("2.4")
+@Name("Indices of List")
+@Description({
+	"Returns all the indices of a list variable, optionally sorted by their values",
+	"To sort the indices, all objects in the list must be comparable;",
+	"Otherwise, this expression will just return the unsorted indices."
+})
+@Examples({"set {l::*} to \"some\", \"cool\" and \"values\"",
+		"broadcast \"%indices of {l::*}%\" # result is 1, 2 and 3", "",
+		"set {_leader-board::first} to 17",
+		"set {_leader-board::third} to 30",
+		"set {_leader-board::second} to 25",
+		"set {_leader-board::fourth} to 42",
+		"set {_ascending-indices::*} to sorted indices of {_leader-board::*} in ascending order",
+		"broadcast \"%{_ascending-indices::*}%\" #result is first, second, third, fourth",
+		"set {_descending-indices::*} to sorted indices of {_leader-board::*} in descending order",
+		"broadcast \"%{_descending-indices::*}%\" #result is fourth, third, second, first"
+})
+@Since("2.4 (indices), INSERT VERSION (sorting)")
 public class ExprIndices extends SimpleExpression<String> {
-	
+
 	static {
 		Skript.registerExpression(ExprIndices.class, String.class, ExpressionType.COMBINED,
-				"[the] (indexes|indices) of %objects%",
-				"(all of the|all the|all) (indices|indexes) of %objects%"
+				"[the] (indexes|indices) of %~objects%",
+				"[the] %~objects%'[s] (indexes|indices)",
+				"[sorted] (indices|indexes) of %~objects% in (ascending|1¦descending) order",
+				"[sorted] %~objects%'[s] (indices|indexes) in (ascending|1¦descending) order"
 		);
 	}
-	
-	@SuppressWarnings("null")
+
+	@SuppressWarnings({"null", "NotNullFieldNotInitialized"})
 	private Variable<?> list;
-	
+
+	private boolean sort;
+	private boolean descending;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		sort = matchedPattern > 1;
+		descending = parseResult.mark == 1;
+		if (exprs[0] instanceof Variable<?>) {
+			list = (Variable<?>) LiteralUtils.defendExpression(exprs[0]);
+
+			if (!list.isList()) {
+				Skript.error("The indices expression must only be used with list variables.");
```

Comment:
```suggestion
				Skript.error("The indices expression must only be used with list variables");
```
same for other error messages